spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -0,0 +1,163 @@
+import org.springframework.cloud.gcp.data.datastore.core.convert.ObjectToKeyFactory;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
+import org.springframework.util.Assert;
+
+/**
+ * An implementation of {@link DatastoreOperations}.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class DatastoreTemplate implements DatastoreOperations {
+
+	private final Datastore datastore;
+
+	private final DatastoreEntityConverter datastoreEntityConverter;
+
+	private final DatastoreMappingContext datastoreMappingContext;
+
+	private final ObjectToKeyFactory objectToKeyFactory;
+
+	public DatastoreTemplate(Datastore datastore,
+			DatastoreEntityConverter datastoreEntityConverter,
+			DatastoreMappingContext datastoreMappingContext,
+			ObjectToKeyFactory objectToKeyFactory) {
+		Assert.notNull(datastore, "A non-null Datastore service object is required.");
+		Assert.notNull(datastoreEntityConverter,
+				"A non-null DatastoreEntityConverter is required.");
+		Assert.notNull(datastoreMappingContext,
+				"A non-null DatastoreMappingContext is required.");
+		Assert.notNull(objectToKeyFactory,
+				"A non-null Object to Key factory is required.");
+		this.datastore = datastore;
+		this.datastoreEntityConverter = datastoreEntityConverter;
+		this.datastoreMappingContext = datastoreMappingContext;
+		this.objectToKeyFactory = objectToKeyFactory;
+	}
+
+	@Override
+	public <T> T findById(Object id, Class<T> entityClass) {
+		return this.datastoreEntityConverter.read(entityClass,
+				this.datastore.get(getKeyFromId(id, entityClass)));
+	}
+
+	@Override
+	public <T> void save(T instance) {
+		this.datastore.put(convertToEntity(instance));
+	}
+
+	@Override
+	public <T> void deleteById(Object id, Class<T> entityClass) {
+		this.datastore.delete(getKeyFromId(id, entityClass));
+	}
+
+	@Override
+	public <T> void delete(T entity) {
+		this.datastore.delete(getKey(entity));
+	}
+
+	@Override
+	public void deleteAll(Class<?> entityClass) {
+		this.datastore.delete(findAllKeys(entityClass));
+	}
+
+	@Override
+	public long count(Class<?> entityClass) {
+		return findAllKeys(entityClass).length;
+	}
+
+	@Override
+	public <T> Collection<T> findAllById(Iterable<?> ids, Class<T> entityClass) {
+		List<Key> keysToFind = new ArrayList<>();
+		ids.forEach(x -> keysToFind.add(getKeyFromId(x, entityClass)));
+		return convertEntities(
+				this.datastore.get(keysToFind.toArray(new Key[keysToFind.size()])),
+				entityClass);
+	}
+
+	@Override
+	public <T> Collection<T> findAll(Class<T> entityClass) {
+		return convertEntities(
+				this.datastore.run(Query.newEntityQueryBuilder()
+						.setKind(this.datastoreMappingContext
+								.getPersistentEntity(entityClass).kindName())
+						.build()),
+				entityClass);
+	}
+
+	@Override
+	public <T> boolean existsById(Object id, Class<T> entityClass) {
+		return findById(id, entityClass) != null;
```

Comment:
Since datastore doesn't seem to support direct `existsById` (or does it), do we really need this method?

Reply:
I think so. The other spring data modules have this in the Operations interface. Also, this is needed by the Repository interfaces that we must implement.

Reply:
OK.